// revportal/settings.js
/*
* Copyright (c) 2014, Rev Software, Inc.
* All Rights Reserved.
*
* This code is confidential and proprietary to Rev Software, Inc
* and may only be used under a license from Rev Software Inc.
*
* Author: <Latheef Shaik>
*/

//for device browser cron log directory path
exports.deviceBrowsersLog = {
    path:"/opt/revportal/log/deviceBrowsersErr.log"
}

//for heatmap cron log directory path
module.exports.heatMapLog = {
    path:"/opt/revportal/log/heatmapErr.log"
};

//for ssl certificates
module.exports.ssl_certificates = {
        is_https:true,
        key:"/etc/ssl/revsw/wildcard.revsw.net.key",
        cert:"/etc/ssl/revsw/wildcard.revsw.net.crt",
        ca:"/etc/ssl/revsw/globalsign_ca.crt"

};

//for mongo constants
module.exports.mongo = {
        url: "54.210.116.53",
        //url: "54.183.7.30",
        port:"27017",
        database:"cube_development",
        is_auth_required: false,
        username: "revuser_prod",
        password: "1QaZTech135",
        event_collection:'pl_info_events'
};

//for forgot password
module.exports.forgot_pwd = {
        url:"https://portal-qa.revsw.net/forgot_pwd.html",
        validate_time:10,
};

// for query details
module.exports.cube_settings = {
        // Network time
        nw_time : "1.0/metric?expression=median(pl_info(networkTime))",
        // BrowserTime
        br_time : "1.0/metric?expression=median(pl_info(browserTime))",
        // BackendTime
        bd_time : "1.0/metric?expression=median(pl_info(backendTime))",
        // PageLoadTime
        page_load_time : "1.0/metric?expression=median(pl_info(pageloadTime))",
        
        step :{ 
                "10sec": "1e4",
                "1min" : "6e4",
                "5min" : "3e5",
                "15min": "9e5",
                "1hr"  : "36e5",
                "1day" : "864e5"
              }
};
